src/fable-library/List.fs

```diff
@@ -6,153 +6,185 @@ module List
+    member xs.Item with get(index) = xs.Values.[xs.Count - index - 1]
+
+let newList (values: ResizeArray<'T>) = { Count = values.Count; Values = values }
+
+let length xs = xs.Count
+
+let empty () = List.Empty
+
+let isEmpty (xs: 'T list) = xs.Count <= 0
+
+let cons (x: 'T) (xs: 'T list) =
+    let values =
+        if xs.Count = xs.Values.Count
+        then xs.Values
+        else xs.Values.GetRange(0, xs.Count)
+    values.Add(x)
+    newList values
+
+let copy (xs: 'T list) =
+    newList <| xs.Values.GetRange(0, xs.Count)
+
+let singleton x =
+    cons x List.Empty
+
+let head (xs: 'T list) =
+    if xs.Count > 0
+    then xs.Values.[xs.Count - 1]
+    else failwith msgListWasEmpty
+
+let tryHead (xs: 'T list) =
+    if xs.Count > 0
+    then Some xs.Values.[xs.Count - 1]
+    else None
+
+let tail (xs: 'T list) =
+    if xs.Count > 0
+    then { Count = xs.Count - 1; Values = xs.Values }
+    else failwith msgListWasEmpty
+
+let last (xs: 'T list) =
+    if xs.Count > 0
+    then xs.Values.[0]
+    else failwith msgListWasEmpty
+
+let tryLast (xs: 'T list) =
+    if xs.Count > 0
+    then Some xs.Values.[0]
+    else None
 
 let compareWith (comparer: 'T -> 'T -> int) (xs: 'T list) (ys: 'T list): int =
     if obj.ReferenceEquals(xs, ys)
     then 0
     else
-        let rec loop xs ys =
-            match xs, ys with
-            | [], [] -> 0
-            | [], _ -> -1
-            | _, [] -> 1
-            | x::xs, y::ys ->
-                match comparer x y with
-                | 0 -> loop xs ys
-                | res -> res
-        loop xs ys
-
-let rec foldIndexedAux f i acc = function
-    | [] -> acc
-    | x::xs -> foldIndexedAux f (i+1) (f i acc x) xs
-
-let foldIndexed<'a,'acc> f (state: 'acc) (xs: 'a list) =
-    foldIndexedAux f 0 state xs
-
-let rec fold<'a,'acc> f (state: 'acc) (xs: 'a list) =
-    match xs with
-    | [] -> state
-    | h::t -> fold f (f state h) t
-
-let reverse xs =
-    fold (fun acc x -> x::acc) [] xs
-
-let foldBack<'a,'acc> f (xs: 'a list) (state: 'acc) =
-    fold (fun acc x -> f x acc) state (reverse xs)
+
+    if isEmpty xs then
+        if isEmpty ys then 0 else -1
+    elif isEmpty ys then
+        1
+    else
+        let mutable i = xs.Count - 1
+        let mutable result = 0
+        if xs.Count > ys.Count then 1
+        elif xs.Count < ys.Count then -1
+        else
+            while i >= 0 && result = 0 do
+                result <- comparer xs.Values.[i] ys.Values.[i]
+                i <- i - 1
+            result
+
+let fold (folder: 'acc -> 'T -> 'acc) (state: 'acc) (xs: 'T list) =
+    let mutable acc = state
+    for i = xs.Count - 1 downto 0 do
+        acc <- folder acc xs.Values.[i]
+    acc
+
+let foldBack (folder: 'T -> 'acc -> 'acc) (xs: 'T list) (state: 'acc) =
+    let mutable acc = state
+    for i = 0 to xs.Count - 1 do
+        acc <- folder xs.Values.[i] acc
+    acc
+
+let reverse (xs: 'a list) =
+    fold (fun acc x -> cons x acc) List.Empty xs
```

Comment:
@alfonsogarciacaro Implemented the reverse optimization. It doesn't seem to have huge performance benefits (although it will save some allocations), most probably my benchmark (FCS) is not bottlenecked by it.